### PR TITLE
Shell add dictionary command

### PR DESCRIPTION
**Pull request by jakub-uC**
Added dictionary commands concept. 

Example usage:

```
static int gain_cmd_handler(const struct shell *shell, size_t argc, char **argv,                                         
                            void *data)                                                                                  
{                                                                                                                        
        int gain;                                                                                                        
                                                                                                                         
        /* data is a value corresponding to called command syntax */                                                     
        gain = (int)data;                                                                                                
                                                                                                                         
        adc_gain_set(gain);                                                                                              
                                                                                                                         
        shell_print(shell, "ADC gain set to: %s\n"                                                                       
                           "Value send to ADC driver: %d",                                                               
                           argv[0],                                                                                      
                           gain);                                                                                        
                                                                                                                         
        return 0;                                                                                                        
}                                                                                                                        
                                                                                                                         
SHELL_SUBCMD_DICT_SET_CREATE(sub_gain, gain_cmd_handler,                                                                 
        (gain_1, 1), (gain_2, 2), (gain_1_2, 3), (gain_1_4, 4)                                                           
);                                                                                                                       
SHELL_CMD_REGISTER(gain, &sub_gain, "Set ADC gain", NULL); 
```

![dict_cmd](https://user-images.githubusercontent.com/34125330/95325248-352f9600-08a1-11eb-8b75-900da9a3a591.png)

### CODE REVIEW

**Comment by nashif** [comment]
> Added dictionary commands concept.

can you add some more details about this concept? How about describing this in the documentation as well?

File: include/shell/shell.h [hunk 1]
```diff
@@ -396,6 +412,52 @@ struct shell_static_entry {
...
+ *		return 0;
+ *	}
+
+ *	SHELL_SUBCMD_DICT_SET_CREATE(sub_dict_cmds, my_handler,
+ *		(value_0, 0), (value_1, 1), (value_2, 2), (value_3, 3)
+ *	);
+ *	SHELL_CMD_REGISTER(dictionary, &sub_dict_cmds, NULL, NULL);
+ */
+#define SHELL_SUBCMD_DICT_SET_CREATE(_name_, _handler_, ...)		\
+	FOR_EACH_FIXED_ARG(SHELL_CMD_DICT_HANDLER_CREATE, (;),		\
```
Comment (nordic-krch) [inline include/shell/shell.h:454]: it should rather have empty separator `()` since it is separating functions.

File: include/shell/shell.h [hunk 2]
```diff
@@ -396,6 +412,52 @@ struct shell_static_entry {
...
+}
+
+/* Internal macro used for creating dictionary commands. */
+#define SHELL_CMD_DICT_CREATE(_data_)					\
+	SHELL_CMD_ARG(GET_ARG_N(1, __DEBRACKET _data_), NULL, NULL,	\
+		UTIL_CAT(cmd_dict_, GET_ARG_N(1, __DEBRACKET _data_)), 1, 0)
+
+/**
+ * @brief Initializes shell dictionary commands.
+ *
```
Comment (nordic-krch) [inline include/shell/shell.h:431]: Please elaborate here when dictionary commands can be used.

File: include/shell/shell.h [hunk 3]
```diff
@@ -139,6 +139,22 @@ const struct device *shell_device_lookup(size_t idx,
 typedef int (*shell_cmd_handler)(const struct shell *shell,
 				 size_t argc, char **argv);
 
+/**
+ * @brief Shell command handler prototype.
```
Comment (nordic-krch) [inline include/shell/shell.h:143]: it is rather handler used for dictionary.

File: include/shell/shell.h (hunk 2 above)
Comment (jakub-uC) [reply include/shell/shell.h:431]: I am working on documentation now. Once I will have it I will update the description as well.

**Comment by jakub-uC** [comment]
> > Added dictionary commands concept.
> 
> can you add some more details about this concept? How about describing this in the documentation as well?

I've updated the documentation and I've added some examples.

**Review by nordic-krch — APPROVED** [review]
looks good

**Review by jukkar — APPROVED** [review]
No objections

**Review by nashif — APPROVED** [review]

**Pull request merged**